### ISSUE TITLE
FAILED: Docker within LXD container on ZFS filesystem with overlay storage driver

### ISSUE THREAD

**Issue by atc0005** [atc0005/ansible-role-lxd-testenv#4]
From `/var/log/messages`:

```
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Starting Docker Socket for the API.
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Listening on Docker Socket for the API.
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/docker.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Starting Docker Application Container Engine...
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /init.scope: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-mqueue.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-udev-trigger.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-journald.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-.lxd\x2dmounts.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-.lxc-sys.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/proc-swaps.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/proc-cpuinfo.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dbus.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-update-utmp.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/containerd.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-tty.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-fuse.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/proc-meminfo.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-journal-flush.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/proc-uptime.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-sysctl.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-null.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-full.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-.lxc-proc.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-remount-fs.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/proc-stat.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-lxd.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/system-getty.slice: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/sys-fs-fuse-connections.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-tmpfiles-setup.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/proc-sys-fs-binfmt_misc.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-ptmx.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/rc-local.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/console-getty.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-udevd.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/cron.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-random-seed.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/rsyslog.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-urandom.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/ssh.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/cgroupfs-mount.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/-.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/sys-kernel-debug.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-net-tun.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-zero.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/resolvconf.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/networking.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/dev-random.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/systemd-user-sessions.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/unattended-upgrades.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/ondemand.service: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to reset devices.list on /system.slice/proc-diskstats.mount: Operation not permitted
Jan 31 17:52:30 ansible-ubuntu-docker dockerd[4085]: time="2019-01-31T17:52:30.352506087Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Jan 31 17:52:30 ansible-ubuntu-docker dockerd[4085]: time="2019-01-31T17:52:30.353023835Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Jan 31 17:52:30 ansible-ubuntu-docker dockerd[4085]: time="2019-01-31T17:52:30.353313051Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Jan 31 17:52:30 ansible-ubuntu-docker dockerd[4085]: time="2019-01-31T17:52:30.353599760Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Jan 31 17:52:30 ansible-ubuntu-docker dockerd[4085]: time="2019-01-31T17:52:30.353958913Z" level=warning msg="[graphdriver] WARNING: the overlay storage-driver is deprecated, and will be removed in a future release"
Jan 31 17:52:30 ansible-ubuntu-docker dockerd[4085]: time="2019-01-31T17:52:30.355444619Z" level=error msg="'overlay' not found as a supported filesystem on this host. Please ensure kernel is new enough and has overlay support loaded." storage-driver=overlay
Jan 31 17:52:30 ansible-ubuntu-docker dockerd[4085]: Error starting daemon: error initializing graphdriver: driver not supported
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: Failed to start Docker Application Container Engine.
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: docker.service: Unit entered failed state.
Jan 31 17:52:30 ansible-ubuntu-docker systemd[1]: docker.service: Failed with result 'exit-code'.
```

Relevant bit:

```
'overlay' not found as a supported filesystem on this host. Please ensure kernel is new enough and has overlay support loaded.
```

**Comment by atc0005** [comment]
Found this LXD GitHub issue:

https://github.com/lxc/lxd/issues/2305

where someone reported that the `vfs` storage driver was being chosen as the default, and was slow. When I either explicitly set that storage driver or renamed the `/etc/docker/daemon.json` file that this role deploys the `vfs` driver was sufficient for the Docker daemon to start and images successfully retrieved/decompressed.

I suspect that upgrading LXD to the Ubuntu 16.04 backports repo version (3.0.3 as of this writing) will be sufficient to allow the `overlay` storage option to work as intended.

**Comment by atc0005** [comment]
Perhaps this isn't a LXD 2.0 issue after all. I just saw this note from the issue referenced earlier:

> overlay/overlay2 is not possible if the host is zfs (as documented by docker)

Sure enough, this latest test that I performed was using LXD + ZFS.

**Comment by atc0005** [comment]
Current Docker (18.09) indicates support for the `zfs` storage driver if using zfs-backed host storage is used.

> The btrfs and zfs storage drivers are used if they are the backing filesystem (the filesystem of the host on which Docker is installed). These filesystems allow for advanced options, such as creating “snapshots”, but require more maintenance and setup. Each of these relies on the backing filesystem being configured correctly.

- https://docs.docker.com/storage/storagedriver/select-storage-driver/
- https://docs.docker.com/storage/storagedriver/zfs-driver/

I've updated my test Ubuntu 16.04 system (which is using a ZFS configuration) to the LXD 3.0.3 and am running the current playbooks from the ansible-playbooks "incubator" repo.

I don't expect this will cause the `overlay` storage driver referenced in `defaults/main.yml` to work, but it's worth ruling out.

**Comment by atc0005** [comment]
As expected, using LXD 3.0.x vs 2.0x wasn't by itself sufficient to resolve the problem:

```
Jan 31 18:58:49 ansible-ubuntu-docker dockerd[3704]: time="2019-01-31T18:58:49.006272914Z" level=warning msg="[graphdriver] WARNING: the overlay storage-driver is deprecated, and will be removed in a future release"
Jan 31 18:58:49 ansible-ubuntu-docker dockerd[3704]: time="2019-01-31T18:58:49.008630819Z" level=error msg="'overlay' not found as a supported filesystem on this host. Please ensure kernel is new enough and has overlay support loaded." storage-driver=overlay
Jan 31 18:58:49 ansible-ubuntu-docker dockerd[3704]: Error starting daemon: error initializing graphdriver: driver not supported
Jan 31 18:58:49 ansible-ubuntu-docker systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
Jan 31 18:58:49 ansible-ubuntu-docker systemd[1]: Failed to start Docker Application Container Engine.
Jan 31 18:58:49 ansible-ubuntu-docker systemd[1]: docker.service: Unit entered failed state.
Jan 31 18:58:49 ansible-ubuntu-docker systemd[1]: docker.service: Failed with result 'exit-code'.
```

**Comment by atc0005** [comment]
I set the storage driver to zfs, restarted Docker and still had no such luck:

```json
{
  "storage-driver": "zfs"
}
```
`root@ansible-ubuntu-docker:~# journalctl -u docker | tail -n 20`

```
Jan 31 19:10:21 ansible-ubuntu-docker systemd[1]: docker.service: Unit entered failed state.
Jan 31 19:10:21 ansible-ubuntu-docker systemd[1]: docker.service: Failed with result 'exit-code'.
Jan 31 19:10:23 ansible-ubuntu-docker systemd[1]: docker.service: Service hold-off time over, scheduling restart.
Jan 31 19:10:23 ansible-ubuntu-docker systemd[1]: Stopped Docker Application Container Engine.
Jan 31 19:10:23 ansible-ubuntu-docker systemd[1]: Starting Docker Application Container Engine...
Jan 31 19:10:23 ansible-ubuntu-docker dockerd[270]: time="2019-01-31T19:10:23.329906199Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Jan 31 19:10:23 ansible-ubuntu-docker dockerd[270]: time="2019-01-31T19:10:23.330362569Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Jan 31 19:10:23 ansible-ubuntu-docker dockerd[270]: time="2019-01-31T19:10:23.330704474Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Jan 31 19:10:23 ansible-ubuntu-docker dockerd[270]: time="2019-01-31T19:10:23.330958669Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Jan 31 19:10:23 ansible-ubuntu-docker dockerd[270]: Error starting daemon: error initializing graphdriver: prerequisites for driver not satisfied (wrong filesystem?)
Jan 31 19:10:23 ansible-ubuntu-docker systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
Jan 31 19:10:23 ansible-ubuntu-docker systemd[1]: Failed to start Docker Application Container Engine.
Jan 31 19:10:23 ansible-ubuntu-docker systemd[1]: docker.service: Unit entered failed state.
Jan 31 19:10:23 ansible-ubuntu-docker systemd[1]: docker.service: Failed with result 'exit-code'.
Jan 31 19:10:25 ansible-ubuntu-docker systemd[1]: docker.service: Service hold-off time over, scheduling restart.
Jan 31 19:10:25 ansible-ubuntu-docker systemd[1]: Stopped Docker Application Container Engine.
Jan 31 19:10:25 ansible-ubuntu-docker systemd[1]: docker.service: Start request repeated too quickly.
Jan 31 19:10:25 ansible-ubuntu-docker systemd[1]: Failed to start Docker Application Container Engine.
Jan 31 19:10:25 ansible-ubuntu-docker systemd[1]: docker.service: Unit entered failed state.
Jan 31 19:10:25 ansible-ubuntu-docker systemd[1]: docker.service: Failed with result 'start-limit-hit'.
```

Looks like ZFS is not an option for my use when using this role *unless* we change the default storage drive to `vfs` as noted earlier.

I think I'll go that route for the best compatibility.

**Comment by atc0005** [comment]
Switching to `vfs` on ZFS test system works well enough (at least to get Docker to start normally).

**Comment by atc0005** [comment]
Overriding via playbook to use the `overlay` Docker storage driver on `dir` LXD storage also works.

**Comment by atc0005** [comment]
Going to consider this resolved once #2 is merged.
